tactics/tacticals.ml

```diff
@@ -516,7 +517,10 @@ module New = struct
     let rest =
       Evd.fold_undefined (fun evk evi acc ->
         match is_undefined_up_to_restriction sigma evk with
-        | Some (evk',evi) -> (evk',evi)::acc
+        | Some (evk',evi) when not (Evd.mem origsigma evk)
```

Comment:
You shouldn't use `Evd.mem` if what you want to check is that the evar is undefined in the original evar map. Also, I am slightly concerned by the conceptual complexity of this function, which starts applying many heuristics and relies on those fragile flags. Maybe we should just sit down and seriously ponder the whole evar aliasing mess instead.